src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs

```diff
@@ -181,9 +183,94 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, INamedTypeSymbol e
+                        conditionNode, out conditionLeft, out conditionRight, out isEquals);
+            }
+        }
+
+        private bool TryAnalyzeBinaryExpressionCondition(
+            ISyntaxFactsService syntaxFacts, TBinaryExpressionSyntax condition,
+            out SyntaxNode conditionLeft, out SyntaxNode conditionRight, out bool isEquals)
+        {
+            isEquals = IsEquals(condition);
+            if (!isEquals && !IsNotEquals(condition))
+            {
+                conditionLeft = null;
+                conditionRight = null;
+                return false;
+            }
+            else
+            {
+                syntaxFacts.GetPartsOfBinaryExpression(condition, out conditionLeft, out conditionRight);
+                return true;
+            }
+        }
+
+        private static bool TryAnalyzeInvocationCondition(
+            SyntaxNodeAnalysisContext context, ISyntaxFactsService syntaxFacts, IMethodSymbol referenceEqualsMethod, TInvocationExpression invocation,
+            out SyntaxNode conditionLeft, out SyntaxNode conditionRight, out bool isEquals)
+        {
+            conditionLeft = null;
+            conditionRight = null;
+            isEquals = true;
+
+            var expression = syntaxFacts.GetExpressionOfInvocationExpression(invocation);
+            var nameNode = syntaxFacts.IsIdentifierName(expression)
+                ? expression
+                : syntaxFacts.IsSimpleMemberAccessExpression(expression)
+                    ? syntaxFacts.GetNameOfMemberAccessExpression(expression)
+                    : null;
+
+            if (!syntaxFacts.IsIdentifierName(nameNode))
+            {
+                return false;
+            }
+
+            syntaxFacts.GetNameAndArityOfSimpleName(nameNode, out var name, out _);
+            if (!syntaxFacts.StringComparer.Equals(name, nameof(ReferenceEquals)))
+            {
+                return false;
+            }
+
+            var arguments = syntaxFacts.GetArgumentsOfInvocationExpression(invocation);
+            if (arguments.Count != 2)
+            {
+                return false;
+            }
+
+            var semanticModel = context.SemanticModel;
+            var cancellationToken = context.CancellationToken;
+            var symbol = semanticModel.GetSymbolInfo(invocation, cancellationToken).Symbol;
+            if (!referenceEqualsMethod.Equals(symbol))
+            {
+                return false;
+            }
+
+            conditionLeft = syntaxFacts.GetExpressionOfArgument(arguments[0]);
+            conditionRight = syntaxFacts.GetExpressionOfArgument(arguments[1]);
+            return true;
+        }
+
         internal static SyntaxNode GetWhenPartMatch(
             ISyntaxFactsService syntaxFacts, ISemanticFactsService semanticFacts, SemanticModel semanticModel, SyntaxNode expressionToMatch, SyntaxNode whenPart)
         {
+            expressionToMatch = GetExpressionIfArgument(syntaxFacts, expressionToMatch);
```

Comment:
I'm curious why this is needed.  The only time you deal with argumetns is back in TryAnalyzeInvocationCondition.  But at the end of that, you always unwrap arguments to expressions.  So it's unclear how you would get an argument here.

Reply:
The code fix calls into this and passes the node it finds at the location given as part of the diagnostic. With the ReferenceEquals case that node is an argument.

Reply:
Ah, that's a result of an earlier piece of code that needs to change.  Specifically, in the fixer change it from:

```c#
                var conditionalExpression = root.FindNode(diagnostic.AdditionalLocations[0].SourceSpan, getInnermostNodeForTie: true);
                var conditionalPart = root.FindNode(diagnostic.AdditionalLocations[1].SourceSpan);
                var whenPart = root.FindNode(diagnostic.AdditionalLocations[2].SourceSpan);
```

To:

```c#
                var conditionalExpression = root.FindNode(diagnostic.AdditionalLocations[0].SourceSpan, getInnermostNodeForTie: true);
                var conditionalPart = root.FindNode(diagnostic.AdditionalLocations[1].SourceSpan, getInnermostNodeForTie: true);
                var whenPart = root.FindNode(diagnostic.AdditionalLocations[2].SourceSpan, getInnermostNodeForTie: true);
```

Then you should not need to do the argument unwrapping.

Reply:
Sorry i didn't catch that earlier!

Reply:
This change seems to work perfectly, thanks @CyrusNajmabadi.